xayn-ai/src/coi/config.rs

```diff
@@ -1,37 +1,125 @@
+        }
+    }
+
     /// The minimum distance between distinct cois.
-    pub threshold: f32,
+    pub(crate) fn threshold(&self) -> f32 {
+        self.threshold
+    }
+
+    /// Sets the threshold for non-negative values.
+    #[cfg(test)]
+    pub(crate) fn with_threshold(self, threshold: f32) -> Result<Self, CoiError> {
+        if threshold >= 0. {
+            Ok(Self { threshold, ..self })
+        } else {
+            Err(CoiError::InvalidThreshold(threshold))
+        }
+    }
+
     /// The positive number of neighbors for the k-nearest-neighbors distance.
-    pub neighbors: NonZeroUsize,
-    /// The time since the last view after which a coi becomes irrelevant.
-    #[allow(dead_code)]
-    pub horizon: Duration,
-    /// The maximum number of key phrases picked during the coi key phrase selection. A coi may have
-    /// more key phrases than this, eg because of merging.
+    pub(crate) fn neighbors(&self) -> usize {
+        self.neighbors
+    }
+
+    /// Sets the neighbors for positive values.
     #[allow(dead_code)]
-    pub max_key_phrases: usize,
+    pub(crate) fn with_neighbors(self, neighbors: usize) -> Result<Self, CoiError> {
+        if neighbors > 0 {
+            Ok(Self { neighbors, ..self })
+        } else {
+            Err(CoiError::InvalidNeighbors(neighbors))
+        }
+    }
+
+    /// The time since the last view after which a coi becomes irrelevant.
+    #[cfg(test)]
+    pub(crate) fn horizon(&self) -> Duration {
+        self.horizon
+    }
+
+    /// Sets the horizon.
+    #[cfg(test)]
+    pub(crate) fn with_horizon(self, horizon: Duration) -> Self {
+        Self { horizon, ..self }
+    }
+
     /// The weighting between coi and pairwise candidate similarites in the key phrase selection.
+    pub(crate) fn gamma(&self) -> f32 {
+        self.gamma
+    }
+
+    /// Sets the gamma for values from the unit interval.
     #[allow(dead_code)]
-    pub gamma: f32,
+    pub(crate) fn with_gamma(self, gamma: f32) -> Result<Self, CoiError> {
+        if (0. ..=1.).contains(&gamma) {
+            Ok(Self { gamma, ..self })
+        } else {
+            Err(CoiError::InvalidGamma(gamma))
+        }
+    }
+
     /// The penalty for less relevant key phrases of a coi in increasing order (ie. lowest penalty
     /// for the most relevant key phrase first and highest penalty for the least relevant key phrase
-    /// last).
-    pub penalty: Vec<f32>,
+    /// last). The length of the penalty also serves as the maximum number of key phrases.
+    #[cfg(test)]
+    pub(crate) fn penalty(&self) -> &[f32] {
+        self.penalty.as_slice()
+    }
+
+    /// Sets the penalty for non-empty, finite values.
+    #[allow(dead_code)]
+    pub(crate) fn with_penalty(self, penalty: &[f32]) -> Result<Self, CoiError> {
+        let penalty = penalty.to_vec();
+        if !penalty.is_empty() && penalty.iter().copied().all(f32::is_finite) {
+            Ok(Self { penalty, ..self })
```

Comment:
in general i agree, but in this case we call the allocation anyways in the success and the error case (the error holds this vec as well), so we don't save any allocation by calling it later.

Reply:
good point, is there any reason why we store the vector in the error?

I personally think of putting anything besides contextual data into an error as (normally) being an
anit-pattern. It looks right, but is in practice tends to be not very useful and can in some cases cause problems (e.g. display formatting too long output). (Through this might not be true in this case as
the penalty size shouldn't be too large as far as I understand.)

Reply:
no particular reason other than having a bit more error info. i think it should also be ok to remove the data in the config errors as the error type & display should be mostly enough to handle the error.

should we simplify the error before merging?